Add iOS test and demo for insecure encryption modes (ECB) (#3547)
OWASP/owasp-mstg: demos/ios/MASVS-CRYPTO/MASTG-DEMO-0068/MASTG-DEMO-0068.md

---
platform: ios
title: Uses of Broken Encryption Modes in CommonCrypto with r2
code: [swift]
id: MASTG-DEMO-0068

> test: MASTG-TEST-0304

Reviewer: Make a similar change to Test ID to `MASTG-TEST-XXXX` to avoid any interference with other pull requests.

```suggestion
test: MASTG-TEST-XXXX
```

    const void *key,
    size_t keyLength,
    const void *iv,         /* optional initialization vector */
    const void *dataIn,     /* optional per op and alg */
    size_t dataInLength,
    void *dataOut,          /* data RETURNED here */
    size_t dataOutAvailable,
    size_t *dataOutMoved);
```

There you will also find the `options` parameter, which can include `kCCOptionECBMode`:

```c
/*!
    @enum        CCOptions
    @abstract    Options flags, passed to CCCryptorCreate().

    @constant    kCCOptionPKCS7Padding   Perform PKCS7 padding.
    @constant    kCCOptionECBMode        Electronic Code Book Mode.
                                         Default is CBC.
*/
enum {
    kCCOptionPKCS7Padding   = 0x0001,
    kCCOptionECBMode        = 0x0002,
};
typedef uint32_t CCOptions;
```

With this information, we can now inspect the disassembled code and, in particular, the third argument of the `CCCrypt` function (`w2`), which has a numeric value `3`. Since there is no enum value `3`, we derive that both `kCCOptionPKCS7Padding` with value `1` and ECB mode option (`kCCOptionECBMode`) with value `2` are set, as `1+2=3`.

> Therefore, the `CCCrypt` function is called with the ECB mode option, AES algorithm, and a key of 16 bytes:

Reviewer: fixed the new line

```suggestion
With this information, you can now examine the disassembled code, focusing specifically on the third argument of the `CCCrypt` function (`w2`), which holds a numeric value of `3`. Given that there is no enum value `3`, you can conclude that both `kCCOptionPKCS7Padding` with a value of `1` and the ECB mode option (`kCCOptionECBMode`) with a value of `2` are enabled, since `1+2=3`. Therefore, the `CCCrypt` function is invoked with the ECB mode option, AES algorithm, and a key that is 16 bytes long.
```

> 

Reviewer: The `evaluation.txt` file is missing. Kindly refer to the example : https://github.com/OWASP/mastg/blob/master/demos/ios/MASVS-CRYPTO/MASTG-DEMO-0018/evaluation.txt
```suggestion
{{ evaluation.txt }}
```
